File: dotnet/samples/SimpleOrm.Sample/Models/UserRole.cs
namespace SimpleOrm.Sample.Models;

/// <summary>
/// Table <c>user_roles</c> (STRICT). Composite key (<c>user_id</c>, <c>role_id</c>),
/// neither part database-generated — the fixture for composite-key support
/// (CLAUDE.md §7.4). Relationships are FK columns only at Level 1; navigation
/// properties are Level 2. <c>CreatedAtUtc</c> from the base doubles as the
/// assignment timestamp.
/// </summary>
[Table("user_roles")]
public sealed class UserRole : BaseModel
{
    [Key]
    [Column]
    [ForeignKey(typeof(User))]
    public long UserId { get; set; }

    [Key]
    [Column]
    [ForeignKey(typeof(Role))]
    public long RoleId { get; set; }

    /// <summary>Populated only by the library (Level 2 loading); no public setter, so it can never disagree with <see cref="UserId"/>.</summary>
    [ManyToOne(nameof(UserId))]
    public User? User { get; private set; }

    /// <summary>Populated only by the library (Level 2 loading); no public setter, so it can never disagree with <see cref="RoleId"/>.</summary>
    [ManyToOne(nameof(RoleId))]
    public Role? Role { get; private set; }
}

File: php/tests/Sample/Models/UserRole.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Tests\Sample\Models;

use SimpleOrm\Metadata\Attributes\Column;
use SimpleOrm\Metadata\Attributes\ForeignKey;
use SimpleOrm\Metadata\Attributes\Index;
use SimpleOrm\Metadata\Attributes\Key;
use SimpleOrm\Metadata\Attributes\ManyToOne;
use SimpleOrm\Metadata\Attributes\Table;

/**
 * Table `user_roles`: composite natural key (`user_id`, `role_id`) — the
 * composite-key fixture (§7.4) and the many-to-many link between users and roles.
 */
#[Table('user_roles')]
#[Index(['roleId'])]
final class UserRole extends BaseModel
{
    #[Key]
    #[Column]
    #[ForeignKey(User::class)]
    public int $userId;

    #[Key]
    #[Column]
    #[ForeignKey(Role::class)]
    public int $roleId;

    /** Added by migration V0005. */
    #[Column]
    public ?string $grantedBy = null;

    #[ManyToOne('userId')]
    public private(set) ?User $user = null;

    #[ManyToOne('roleId')]
    public private(set) ?Role $role = null;
}
